frb/analysis/dmhalos_photoz.py

```diff
@@ -0,0 +1,615 @@
+
+    return mhalo_reals
+
+def _dm_pdf(cigale_tab:Table, eazy_outdir:str,
+            mean_interp:interp2d, stddev_interp:interp2d,
+            ang_dia_interp:interp1d, dm_interpolator:RegularGridInterpolator,
+            n_cores:int = 8):
+    """
+    For a given galaxy, compute its PDF of
+    DM from the CIGALE and EAZY inputs.
+    Args:
+        cigale_tab (Table): On of the groups
+            from the full cigale result. This
+            group contains data on only one galaxy
+            at various assumed redshifts. 
+        eazy_outdir (str): Path to the directory with EAZY output
+        mean_interp (interp2d): <log_mhalo(log_mstar, z)> (based on SHMR)
+        stddev_interp (interp2d): std.dev. log_mhalo(log_mstar, z) (based on SHMR)
+        ang_dia_interp (interp1d): angular_diameter_distance(z) (Planck15)
+        dm_interpolator (RegularGridInterpolator): DM(z, offset_kpc, log_mhalo)
+        n_cores (int, optional): Number of CPU threads to use.
+    Returns:
+        dm_values (np.ndarray): Array containing DM realizations for the galaxy.
+        z_draws (np.ndarray): Array containing redshift draws from which dm_values were produced.
+    """
+    
+    # Prepare interpolation functions from the
+    # CIGALE table
+    log_mstar_interp = interp1d(cigale_tab['redshift'], cigale_tab['log_mstar'], bounds_error=False, fill_value=1)
+    log_mstar_err_interp = interp1d(cigale_tab['redshift'], cigale_tab['log_mstar_err'], bounds_error=False, fill_value=1)
+
+    # Get 1000 random redshift draws from EAZY
+    z_draws = _sample_eazy_redshifts(cigale_tab['gal_ID'][0], eazy_outdir)
+    if np.isscalar(z_draws):
+        return -99.
+
+    # Convert the photo-z draws to mean stellar masses and errors
+    log_mstar_array = log_mstar_interp(z_draws)
+    log_mstar_err_array = log_mstar_err_interp(z_draws)
+
+    func = lambda idx: _mhalo_realizations(log_mstar_array[idx], log_mstar_err_array[idx], z_draws[idx], mean_interp, stddev_interp)
+
+    # Draw stellar mass values from a normal distribution and produce halo
+    # masses, halo_mass errors
+    p = Pool(n_cores)
+    log_mhalos = p.map(func, np.arange(len(z_draws)))
+    zz_draws = np.repeat(z_draws, len(log_mhalos[0]))
+    offsets = ang_dia_interp(z_draws)*cigale_tab['sep_ang'][0]*u.arcmin.to('rad')
+    oo_draws = np.repeat(offsets, len(log_mhalos[0]))
+    dm_values = dm_interpolator((zz_draws, oo_draws, np.concatenate(log_mhalos)))
+
+    return dm_values, z_draws.astype('float32')
+
+def dm_grid(frb_z:float, outdir:str=DEFAULT_DATA_FOLDER, outfile:str=None)->None:
+    """
+    Produce DM estimates for a 3D grid of
+    redshift, offsets and log_halo_masses and write
+    them to disk.
+    Args:
+        frb_z(float): frb redshift
+        outdir(str, optional): data directory to store results
+        outfile(str, optional): name of results fits file (within outdir).
+    """
+    # Redshift grid
+    n_z = 100
```

Comment:
I think these should parameters to the method.

Reply:
Changed.